### PR TITLE
Allow exception-patterns under or-patterns

### PR DESCRIPTION
**Pull request by trefis**
Revival of #305 .

A few things of note compared to the previous version:
- `Parmatch` and `Matching` always `assert false` when they see a `Tpat_exception`
- `Parmatch.pressure_variants` is called on the value and exception patterns independently.
  This wasn't done in the previous version and was the cause of [PR#7083](https://caml.inria.fr/mantis/view.php?id=7083).
- `Parmatch.check_ambiguous_bindings` is called independently on the value and exception patterns. As for the other operations in parmatch, it wouldn't make sense to mix both kind of patterns in a single matrix.
- In `Rec_check` I believe I handle cases containing only exception-patterns in exactly the same way as before. However, I wonder if perhaps this special casing could not simply be dropped (considering that `Tpat_exception` is not a destructuring pattern). I didn't try to understand how the check works though (nor do I plan to right now) so I'm going to leave things as is unless @yallop says it can be simplified.
- in `Translcore` I make sure to remember (i.e. save in `try_ids`) all the names given to the exception of an exception pattern so that reraise works properly. I also added a test to make sure reraise works as intended when raising from a branch when we either match or catch an exception.

Apart from that everything should be as in #305, in particular : mixing value and exception patterns under a guard is still disallowed.

---

Off-topic:
- I noticed that `Parmatch.check_ambiguous_bindings` is delayed when the matching contains polymorphic variants (and I didn't change it here) but I don't think that is necessary.
- ~This PR is based on top of #1567 , so the first 3 commits can be disregarded.~

### CODE REVIEW

File: bytecomp/matching.ml [hunk 1]
```diff
@@ -2824,6 +2825,7 @@ let find_in_pat pred =
         find_rec p || find_rec q
     | Tpat_constant _ | Tpat_var _
     | Tpat_any | Tpat_variant (_,None,_) -> false
+  | Tpat_exception _ -> assert false
```
Comment (kit-ty-kate) [inline bytecomp/matching.ml]: Is there an indentation issue or did it mean something else ?

Comment (trefis) [reply bytecomp/matching.ml]: Looks like indentation/copy-paste failure on my part; it's now fixed.

Thanks!

**Comment by trefis** [comment]
Rebased, as #1567 is now on top of #1573, the first 5 commits can be ignored.
The tests should now pass on both linux and windows!

**Comment by trefis** [comment]
I just rebased this PR on trunk now that the #1573 and #1567 have been merged. This should therefore be much easier to review.

Also, it should be noted that `js_of_ocaml` has been updated to handle exceptions under or-patterns!

**Comment by trefis** [comment]
I rebased this once more!

I also pinged @gasche by email to get some review, who (for obscure reasons) said it wasn't reasonable (sic) for him to review this.
So let me repeat some of my earlier points and ping some people explicitly:
- [testsuite/tests/match-exception-warnings/placement.ml](https://github.com/ocaml/ocaml/pull/1568/files#diff-cea8272d585f841d08ab92898ed45f43) tests all the syntactically valid patterns involving a `Ppat_exception`.
In particular one can see that exception patterns are now allowed under: or-patterns, local opens, type constraints. And that's the only change, every other appearance is still rejected.
- [`Rec_check`](https://github.com/ocaml/ocaml/pull/1568/files#diff-03276fb19e18b9adb70813d03c1c185a) has been updated.
I suspect that the changes are correct but could be simplified. I would very much appreciate a review from @yallop on this.
- `Parmatch` will error out (`assert false` or `fatal_error`) if it ever sees a `Tpat_exception`: there is no real need for review here (but the callers must be checked).
- The compilation (most of the changes happening in [`Translcore`](https://github.com/ocaml/ocaml/pull/1568/files#diff-e8899086c6b3526d6bc4453b4ad5395c)) is as it was in #305 .
I still believe it is fine, but a fresh review from @chambart or @maranget can't hurt.
- The changes in [`Typecore`](https://github.com/ocaml/ocaml/pull/1568/files#diff-ff97e134318d4be6e32d0a612a42cf48) are the ones that need review the most: they aren't that big, but that's where the issue was missed last time. The changes to `type_pat` are fairly trivial, but a complete review `type_cases` is required.
@garrigue , @lpw25: would one of you mind giving it a look?

**Comment by yallop** [comment]
> In `Rec_check` I believe I handle cases containing only exception-patterns in exactly the same way as before. However, I wonder if perhaps this special casing could not simply be dropped (considering that `Tpat_exception` is not a destructuring pattern). I didn't try to understand how the check works though (nor do I plan to right now) so I'm going to leave things as is unless @yallop says it can be simplified.
[...]
> I suspect that the changes are correct but could be simplified.

Yes -- it looks to me as though `is_exception_pattern` could be dropped altogether, leaving the code in `case` as it was previously.

File: testsuite/tests/backtrace/backtrace_or_exception.ml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+(* TEST
+   flags = "-g"
+   ocamlrunparam += ",b=1"
+   * bytecode
+     reference = "${test_source_directory}/backtrace_or_exception.byte.reference"
```
Comment (dra27) [inline testsuite/tests/backtrace/backtrace_or_exception.ml]: This line is over 80 chars!

**Review by dra27 — DISMISSED** [review]
Excuse the nit, which is #1148-related

File: testsuite/tests/backtrace/backtrace_or_exception.ml (hunk 2 above)
Comment (trefis) [reply testsuite/tests/backtrace/backtrace_or_exception.ml]: Not anymore! :)

**Comment by trefis** [comment]
@yallop: thanks! I pushed a commit doing the simplification.

File: testsuite/tests/backtrace/backtrace_or_exception.ml (hunk 2 above)
Comment (dra27) [reply testsuite/tests/backtrace/backtrace_or_exception.ml]: I thank you - now that the an off-by-one error in the Travis CI script is fixed, it'll even get scanned next time 🙂

**Review by lpw25 — APPROVED** [review]
I've reviewed the type-checking stuff and it looks fine. Since the other parts have already been reviewed, I think this is ready to merge.

**Comment by trefis** [comment]
Thanks for the review!

I'll rebase and merge this on Monday, unless someone objects.

**Pull request merged**